### ISSUE TITLE
[CPU] Stack allocation limit exceeded when compiling T5-Large with microkernels enabled

### ISSUE THREAD

**Issue by mariecwhite** [iree-org/iree#14414]
### What happened?

Error when compiling some models with `--iree-flow-enable-data-tiling --iree-llvmcpu-enable-microkernels`:

```
2023-07-17T00:21:19.4435099Z [149/288] Generating /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_T5_LARGE_FP32_JAX_512XI32_BATCH1_stablehlo___x86_64-cascadelake-linux_gnu-llvm_cpu__experimental-flags_microkernels_/module.vmfb from T5_LARGE_FP32_JAX_512XI32_BATCH1(stablehlo) [x86_64-cascadelake-linux_gnu-llvm_cpu][experimental-flags,microkernels]
2023-07-17T00:21:19.4438993Z FAILED: e2e_test_artifacts/iree_module_T5_LARGE_FP32_JAX_512XI32_BATCH1_stablehlo___x86_64-cascadelake-linux_gnu-llvm_cpu__experimental-flags_microkernels_/module.vmfb /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_T5_LARGE_FP32_JAX_512XI32_BATCH1_stablehlo___x86_64-cascadelake-linux_gnu-llvm_cpu__experimental-flags_microkernels_/module.vmfb 
2023-07-17T00:21:19.4443763Z cd /work/build-e2e-test-artifacts/tests/e2e/test_artifacts && /work/full-build-dir/install/bin/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=stablehlo --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu --iree-llvmcpu-target-cpu=cascadelake --iree-flow-enable-data-tiling --iree-llvmcpu-enable-microkernels /work/build-e2e-test-artifacts/e2e_test_artifacts/model_T5_LARGE_FP32_JAX_512XI32_BATCH1.mlirbc -o /work/build-e2e-test-artifacts/e2e_test_artifacts/iree_module_T5_LARGE_FP32_JAX_512XI32_BATCH1_stablehlo___x86_64-cascadelake-linux_gnu-llvm_cpu__experimental-flags_microkernels_/module.vmfb --iree-hal-executable-object-search-path=\"/work/build-e2e-test-artifacts\"
2023-07-17T00:21:19.4447899Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:641:12: error: 'func.func' op exceeded stack allocation limit of 32768 bytes for function. Got 65536 bytes
2023-07-17T00:21:19.4448879Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:2:3: note: called from
2023-07-17T00:21:19.4455188Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:641:12: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
2023-07-17T00:21:19.4460195Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:2:3: note: called from
2023-07-17T00:21:19.4461047Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:641:12: error: failed to serialize executables
2023-07-17T00:21:19.4461936Z /tmp/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlir:2:3: note: called fro
```

### Steps to reproduce your issue

Download https://storage.googleapis.com/iree-model-artifacts/jax/jax_models_0.4.13_1688607404/T5_LARGE_FP32_JAX_512XI32_BATCH1/stablehlo.mlirbc

Run:
```
iree-compile --output-format=vm-bytecode \
    --mlir-print-op-on-diagnostic=false \
    --iree-hal-target-backends=llvm-cpu \
    --iree-input-type=stablehlo \
    --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu \
    --iree-llvmcpu-target-cpu=cascadelake \
    --iree-flow-enable-data-tiling \
    --iree-llvmcpu-enable-microkernels \
    stablehlo.mlirbc -o module.vmfb
```

### What component(s) does this issue relate to?

Compiler

### Version information

ffb5747c9c2f511a96693ee10eb4fcbaafb4d110

**Comment by MaheshRavishankar** [comment]
Cc @bjacob and @hanhanW 

@mariecwhite we saw similar issues recently and Hanhan and Benoit fixed those. Are you on ToT?

**Comment by mariecwhite** [comment]
Looks like I didn't have https://github.com/openxla/iree/commit/09685ee644d2d5676f9566e3275bca566efe547d. Re-running now.

**Comment by mariecwhite** [comment]
Getting the same error after rebasing to ToT. Occurs on T5-Large for both TF and JAX.

**Comment by bjacob** [comment]
I'll look into this today.

**Comment by bjacob** [comment]
Minimized testcase:

```mlir
func.func public @main(%arg0 : tensor<1x512x1024xf32>) -> tensor<1x512x1024xf32> {
  %93 = stablehlo.constant dense<0.0> : tensor<1024x1024xf32>
  %1343 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
  %1344 = stablehlo.reduce(%arg0 init: %1343) applies stablehlo.add across dimensions = [2] : (tensor<1x512x1024xf32>, tensor<f32>) -> tensor<1x512xf32>
  %1345 = stablehlo.broadcast_in_dim %1344, dims = [0, 1] : (tensor<1x512xf32>) -> tensor<1x512x1024xf32>
  %1360 = stablehlo.dot_general %1345, %93, contracting_dims = [2] x [0], precision = [DEFAULT, DEFAULT] : (tensor<1x512x1024xf32>, tensor<1024x1024xf32>) -> tensor<1x512x1024xf32>
  return %1360 : tensor<1x512x1024xf32>
}
```

Reproduction command line is still as in the original repro above:

```
iree-compile --output-format=vm-bytecode \
    --mlir-print-op-on-diagnostic=false \
    --iree-hal-target-backends=llvm-cpu \
    --iree-input-type=stablehlo \
    --iree-llvmcpu-target-triple=x86_64-unknown-linux-gnu \
    --iree-llvmcpu-target-cpu=cascadelake \
    --iree-flow-enable-data-tiling \
    --iree-llvmcpu-enable-microkernels \
    /tmp/a.mlir -o /tmp/a.vmfb
```

**Comment by MaheshRavishankar** [comment]
@mariecwhite to unblock for these benchmarks you can try adding `--iree-llvmcpu-fail-on-out-of-bounds-stack-allocation=false`  and make sure it passes. We can fix the issue asynchronously. The stack limit is a fail-safe that we add to avoid some portability issues. I have this on my list of things

**Comment by hanhanW** [comment]
I plan to switch all experimental-flags benchmarks to data tiling + microkernels after landing my multi-lowering work. At that point, we have to add `--iree-llvmcpu-fail-on-out-of-bounds-stack-allocation=false` flag; I will work on other fixes to remove the flag afterwards. So feel free to add the flag, and pass the issue to me.

**Comment by bjacob** [comment]
Further minimized testcase, down to one dispatch, which has some producer `linalg.generic` folded into the packing into the matmul LHS. Still same repro command line and same error as above.

This needs to have https://github.com/openxla/iree/pull/14421 applied though, otherwise this was running into a separate issue.

```mlir
func.func @dispatch(%5 : tensor<512x1024xf32>, %10 : index, %8 : index)
    -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<512x1024xf32>>> {
  %cst = arith.constant 0.000000e+00 : f32
  %6 = tensor.empty() : tensor<512x1024xf32>
  %7 = tensor.empty() : tensor<512xf32>
  %9 = linalg.fill ins(%cst : f32) outs(%7 : tensor<512xf32>) -> tensor<512xf32>
  %11 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]}
    ins(%5 : tensor<512x1024xf32>) outs(%9 : tensor<512xf32>) {
  ^bb0(%in: f32, %out: f32):
    %14 = arith.addf %out, %in : f32
    linalg.yield %14 : f32
  } -> tensor<512xf32>
  %12 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]}
    ins(%11 : tensor<512xf32>) outs(%6 : tensor<512x1024xf32>) {
  ^bb0(%in: f32, %out: f32):
    linalg.yield %in : f32
  } -> tensor<512x1024xf32>
  %padded = tensor.pad %12 low[0, 0] high[%10, %8] {
  ^bb0(%arg6: index, %arg7: index):
    tensor.yield %cst : f32
  } : tensor<512x1024xf32> to tensor<?x?xf32>
  %13 = iree_linalg_ext.set_encoding %padded : tensor<?x?xf32> -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<512x1024xf32>>>
  return %13 : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<512x1024xf32>>>
}
```

**Comment by bjacob** [comment]
Tentative fix: https://github.com/openxla/iree/pull/14427

**Comment by bjacob** [comment]
This has been fixed by @hanhanW 's #14068. Thanks!

**Comment by hanhanW** [comment]
> This has been fixed by @hanhanW 's #14068. Thanks!

cool, thanks for verifying!
